src/engine/strat_engine/engine.rs

```diff
@@ -124,6 +140,51 @@ impl Engine for StratEngine {
         Ok(uuid)
     }
 
+    /// Evaluate a device node & devicemapper::Device to see if it's a valid
+    /// stratis device.  If all the devices are present in the pool and the pool isn't already
+    /// up and running, it will get setup and the pool uuid will be returned.
+    fn block_evaluate(&mut self,
+                      dev_node: PathBuf,
+                      device: Device)
+                      -> EngineResult<Option<PoolUuid>> {
+        let mut rc = None;
+        let mut duplicate = false;
+
+        if let Some(pool_uuid) = is_stratis_device(&dev_node)? {
+
```

Comment:
Please remove use of ```rc``` and ```duplicate``` variables by restructuring like:

```
if let Some(pool_uuid) = ... {
  if self.pools.contains(uuid) {
    // warn
    // remove pool_uuid entry from incomplete pools
    // Ok(None)
  } else {
    // insert current device into incomplete pools
    // get the devices corresponding to pool_uuid from incomplete pools
    // let rc = if setup of new pool succeeds
    //        remove from incomplete pools
    //        insert into pools
    //        Some(pool_uuid)
    // else
    //        None
    // Ok(rc)
  }
  // no return value here, cause it's taken care of by the if block.
}
```

You have to call remove on the incomplete pools on two different paths, but you get to eliminate two varaibles with semantic meaning.
I think it is less confusing this way.

Reply:
I realized 2 further things about this code while riding the bus.

1. You shouldn't remove the ```pool_uuid``` entry from ```incomplete_pools``` if it is in ```pools```. You can show that the engine always maintains the invariant that the keys of ```pools``` and the keys of ```incomplete_pools``` are disjoint sets. If the key is in one, then it can't be in the other.

If this ever turns out to be false, it would actually be nice to look at some data indicating where you went wrong. So, never any reason to remove from ```incomplete_pools``` in this situation.

2. There is a possibility that when you insert into ```pools``` a pool that is already there will be ejected. This is because ```pools``` actually has two unique ids, the pool name and the pool uuid. Here you ensure that there is no conflict with the uuids, but it is easy to come up with a sequence of events that result in a name conflict. I think that in that case the newly created pool should be held in a special place reserved for problematic pools, i.e, the existing pool should not be ejected.

Reply:
Will calling `setup` with a duplicate pool name, but different uuid cause `setup` to fail?  If not then we should restructure the code, because it's a bad situation where we can succeed at calling `setup` only to fail because we cannot store the pool in the table data structure.  IMHO I'm thinking that we should have one entry point which brings up a pool and inserts it into the correct data structure(s).

Reply:
I'm generally against having methods that work by side-effects if that can be avoided. Also, minimizing the number of parameters is generally a good idea.

You say it's a bad situation, where a pool can be setup, but turn out to be in a name-duplicate. Why is that a problem, and how would the restructuring that you are suggesting solve or alleviate the problem? Note that this situation is also encountered in ```StratPool::initialize```, where the action taken on a name conflict is just to return an error. I suspect that, given the new more forgiving context of handling incomplete pools that this PR introduces, that may be the wrong action to take in that context.

Reply:
I guess, more fundamentally, and regardless of the code, what ought to be done if:
1. You find some blockdevs which are a bad fit for the current world, because they contain the same name as some other blockdevs that are already in an established pool, but a different UUID?
2. What should be done if these odd blockdevs form a complete pool, or should the code stop before it even gets to fidning out if that is true or not?

Reply:
`StratPool::setup` has a side effect that it makes the pool active and available on the system.  I dislike that at the moment we have to call setup, go through that entire process to get the pool object so we can ask the question, what is your name (yes we could add a method for that)?  Then we realize that we have a pool by that name which violates a rule in the Engine, but we already made it public.

I've made a change to my PR so in this case we will `teardown` the newly found pool and leave it in the incomplete pools while logging a error until we figure out what we ideally want to do.  The user could recover by renaming the existing live pool and then generate a synthetic hot plug event on one of the other devices in the other inactive pool.  We would then dynamically bring the pool back up.

Reply:
We haven't made the pool public as a consequence of calling ```setup()```. And the purpose of calling ```setup()``` is to set up the pool, we can't really get around that.

The problem is that the devnode itself doesn't contain the name of the pool it belongs to. That's in the pool's metadata which can only be discovered by reading the blockdevs belonging to the pool and finding the ones that contain the most recent metadata. That's why getting that information is in ```StratPool::setup()```. But, if it were necessary or desirable, it could probably be lifted out of ```StratPool::setup()``` since the whole activity is encapsulated in the ```get_metadata``` method.

Reply:
> We haven't made the pool public as a consequence of calling `setup()`

Are you sure?  It sure looks like after setup dm tables have been loaded and the associated block devices are now present in `/dev`

Reply:
I think this is just a being divided by a common language kind of thing. Of course, we have set up many dm devices, and I agree that that is not what we would wish to do.